feat(aio): add prettier network fail page

wardbell opened the pull request:
closes #16112.

When server or network goes down, we serve a canned page. Current ugly page is described in issue #16112.

This PR provides a prettier page along the lines of the better looking 404 page.
It does NOT cache the page so that, once connectivity is restored, it will try again to get that page.

I found a free “network crashed” svg that’s tiny. I’m loading it in `index.html` for the obvious reason that it needs to exist when you crash.  It should go in the service worker `pwa-manifest.json` too but I don’t know how to do that.

I would have just put it inline but I don’t know how to position and scale an svg. Maybe this is much ado about too little and we should just use a Material Design Icon.

### Update (2nd commit)
- fixed `DocumentService` test I broke (refactored for clarity)
- switched from svg to Material Design icon for lighter weight and fewer complications
- generalized for non-network fetch error causes.

**See it work**

1. Open the Mary Poppins preview
2. Nav to Docs
3. Pick a guide doc
4. Open dev tools, network tab
5. Click "offline"
6. Nav to a different guide page; should see the new error page
7. Nav back to prev. page; it should still display perfectly
8. Go back online
9. Nav to the formerly-failed guide page; it loads this time.

**Todo**
- [ ] @sjtrimble to fix styling.

alfaproject commented on aio/src/app/documents/document.service.ts:
```diff
@@ -15,11 +15,21 @@ import { LocationService } from 'app/shared/location.service';
 export class DocumentService {
 
-  private cache = new Map<string, Observable<DocumentContents>>();
+  private cache  = new Map<string, Observable<DocumentContents>>();
```
revert

mary-poppins commented:
The angular.io preview for daa7d40910fcd57ff0c53b28765ff0c289c5b669 is available [here][1].

[1]: https://pr16139-daa7d40910fcd57ff0c53b28765ff0c289c5b669.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 739d3188567261115810ba7cddf79f7678891751 is available [here][1].

[1]: https://pr16139-739d3188567261115810ba7cddf79f7678891751.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 5d8879e696ed706f5c5d33b328b7b5c78bf8bd27 is available [here][1].

[1]: https://pr16139-5d8879e696ed706f5c5d33b328b7b5c78bf8bd27.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 71fa91d5610a517c4ddad913e4c473ac94d22c47 is available [here][1].

[1]: https://pr16139-71fa91d5610a517c4ddad913e4c473ac94d22c47.ngbuilds.io/

gkalpak reviewed (approved):
LGTM 👍 

<sub>Not related to this PR, but I wish we would use consistent names for variables 😛</sub>

petebacondarwin commented on aio/src/app/documents/document.service.ts:
```diff
@@ -32,16 +42,18 @@ export class DocumentService {
+    if (!this.cache.has(path)) {
+      return this.fetchDocument(path, docId).do(doc => {
+        if (!doc.nocache) { this.cache.set(path, of(doc)); }
+      });
```
Only adding to the cache after the document arrives means that multiple hits to this URL will cause multiple requests until one of them responds.
We should add the observable to the cache when the request is made but then remove it from the cache if the doc that is returned should not be cached.

petebacondarwin replied:
Something like:

```
this.cache.set(path, this.fetchDocument(path).do(doc => {
  if (doc.nocache) { this.cache.remove(path); }
});
```

wardbell replied:
Good idea

IgorMinar commented on aio/src/app/documents/document.service.ts:
```diff
@@ -15,6 +15,16 @@ import { LocationService } from 'app/shared/location.service';
+<div class="nf-icon material-icons">error_outline</div>
+<div class="nf-response l-flex-wrap flex-center">
+<h2>Request for document failed</h2>
+<p>We're sorry but we are unable to retrieve the "{{doc}}" page at this time.
```
comma missing before "but"?

IgorMinar replied:
This text is confusing. How about something shorter and more to the point like: `We are sorry, but due to a network issue we can't load document "{{doc}}". Please check your connection and try again.`

wardbell replied:
Because the error isn't always a network issue. It arises from bad JSON too and who knows what else.

wardbell replied:
We can drop the icon altogether too. I just added one because the 404 page had one.

IgorMinar commented on aio/src/app/documents/document.service.ts:
```diff
@@ -15,6 +15,16 @@ import { LocationService } from 'app/shared/location.service';
 const FILE_NOT_FOUND_URL = 'file-not-found';
+const FAIL_PAGE = `
+<div class="content"><div class="nf-container l-flex-wrap flex-center">
+<div class="nf-icon material-icons">error_outline</div>
```
What's the purpose of "error_outline" text in here? I've seen you use this kind of text elsewhere and it's this text that is causing the flicker. Please remove.

wardbell replied:
I'm using them exactly as prescribed by the [Material Design Guidelines](http://google.github.io/material-design-icons/), e.g., 
```
<i class="material-icons">face</i>
```

In this case, flicker is highly improbably because there is almost no way to get to this page before the font loads.  Therefore, I'm keeping it.

But we need a broader strategy for using MD icons. That should be a separate issue.

IgorMinar replied:
Interesting, I didn't know that MD fonts were using ligatures. In that case we must ensure that the font is loaded before we render anything that depends on it.

IgorMinar commented on aio/src/app/documents/document.service.ts:
```diff
@@ -15,6 +15,16 @@ import { LocationService } from 'app/shared/location.service';
 import { Logger } from 'app/shared/logger.service';
 
 const FILE_NOT_FOUND_URL = 'file-not-found';
+const FAIL_PAGE = `
```
why is this page implemented differently than the file not found page? you could implement in the same way as 404 and then use service worker and/or `<link rel="preload"` to cache it. (https://css-tricks.com/prefetching-preloading-prebrowsing/)

wardbell replied:
Reasons:
1) Service worker is not enabled on most devices.
2) This approach is guaranteed to work even when there is no network (assuming that you've loaded the JS of course).
3) MDN is not clear on browser support. 

Why worry? If we're going to get it anyway, and we inline templates routinely, what's wrong with doing it here? Seems simple and sure to me.

IgorMinar commented:
We also need a different design for this page. This is way too scary and bold.

Among other things the fact that the header is not aligned along with the icon makes it look quite bad.


Current screenshot: 
<img width="1662" alt="screen shot 2017-04-19 at 3 31 48 pm" src="https://cloud.githubusercontent.com/assets/216296/25185125/bc4ff0c4-251c-11e7-904d-511f122ff34a.png">

wardbell commented:
Stef and I will work on the cosmetic issues in a separate PR.

petebacondarwin commented:
> @petebacondarwin Please note that the entry in the cache for a document that 404s now has the url of the page that was sought rather than the url of the 404 page that we present. Revised tests reflect this coincidental change.

No! 😱  That was the whole point of my change previously! We need to be able to assign the `pageId` in the `AppComponent` based on the document being rendered not on the current URL. The test was there on purpose and not incidentally incorrect. I think that calling it URL was a red herring that has confused matters.

petebacondarwin committed:
build(aio): add the `id` to the DocumentContents at doc-gen time

petebacondarwin committed:
refactor(aio): rename DocumentContents `url` to `id` and tidy up

Dgeni is now providing the `id` for all the documents to be viewed. So we
no longer need to add this to the DocumentContents object.

There are some notable changes in the refactoring:

`DocumentService`:

* The id of the document to render is now obtained from `LocationService.path()`.
* The `getFileNotFoundDoc` and `getErrorDoc` methods have been extracted from
the `fetchDocument` method.

`AppComponent`:

* the `pageId` is now computed in a separate `setPageId` method.

`AppComponen` spec file:

* The `TestHttp` has had the hard coded documents removed and replaced with
a function that will generate docs as needed.

petebacondarwin committed:
feat(aio): add `<current-location>` embedded component

wardbell committed:
feat(aio): add prettier network fail page

IgorMinar commented on aio/src/index.html:
```diff
@@ -27,6 +27,9 @@
   <meta name="theme-color" content="#1976d2">
 
+  <!-- loading this image here so we will have it already if the network fails -->
+  <img src="assets/images/support/net-crash.svg" width="0px;" height="0px;">
```
I don't think we use this any more. remove?

petebacondarwin replied:
done

petebacondarwin commented on aio/src/app/app.component.spec.ts:
```diff
@@ -236,11 +236,6 @@ describe('AppComponent', () => {
       expect(component.pageId).toEqual('guide-other');
       expect(container.properties['id']).toEqual('guide-other');
-
-      locationService.go('guide/http#anchor-1');
```
we can't actually test hash removal here because the `MockLocationStrategy` doesn't support stripping them even if you do pass `false` to the `path(false)` method.

petebacondarwin commented on aio/src/app/documents/document.service.ts:
```diff
@@ -28,50 +39,53 @@ export class DocumentService {
     location: LocationService) {
     // Whenever the URL changes we try to get the appropriate doc
-    this.currentDocument = location.currentUrl.switchMap(url => this.getDocument(url));
+    this.currentDocument = location.currentUrl.switchMap(() => this.getDocument(location.path()));
```
`location.path()` now does all the cleaning that we need.

petebacondarwin commented on aio/src/app/shared/location.service.ts:
```diff
@@ -13,6 +13,7 @@ export class LocationService {
 
   private readonly urlParser = document.createElement('a');
   private urlSubject = new Subject<string>();
+  private currentPath: string;
```
oops this is unused!

petebacondarwin committed:
feat(aio): add LocationService.path() method

This enables things like embedded components easy access to a clean version
of the current location path.

petebacondarwin committed:
feat(aio): add prettier network fail page

mary-poppins commented:
The angular.io preview for 23b57aa35faf217d449b11b3deb5bca02160c59c is available [here][1].

[1]: https://pr16139-23b57aa35faf217d449b11b3deb5bca02160c59c.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 1925d49eb219876ae16d29a145fff537cb36e4fb is available [here][1].

[1]: https://pr16139-1925d49eb219876ae16d29a145fff537cb36e4fb.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 9f14ce5218307e2eaab63b290192ead100889fe7 is available [here][1].

[1]: https://pr16139-9f14ce5218307e2eaab63b290192ead100889fe7.ngbuilds.io/

wardbell reviewed (commented):
Looks great, Pete. Ship it!

mhevery commented:
closed with 97303510483c898b0de12a0112398e84e593f064
